File: Source/Obsequy.Web/App/Schemes/providerResponseScheme.js


angular.module('providerResponse-scheme', ['underscore'])
 .factory('ProviderResponse', ['_', 'ResponseAgreement', 'ResponseAlternate', 'ResponseReceipt', 'ConsumerPrincipal', 'ConsumerMember', 'ConsumerPreference', 'ConsumerSchedule', 'ConsumerRequest', function (_, ResponseAgreement, ResponseAlternate, ResponseReceipt, ConsumerPrincipal, ConsumerMember, ConsumerPreference, ConsumerSchedule, ConsumerRequest) {
 	var ProviderResponse = function (properties) {

		angular.extend(this, {

			id: undefined,
			consumerPortfolioId: undefined,
			providerPortfolioId: undefined,

			distance: undefined,
			quote: undefined,
			depositDue: undefined,
			depositPaid: undefined,
			balanceDue: undefined,
			balancePaid: undefined,

			agreement: new ResponseAgreement(),
			alternate: new ResponseAlternate(),

			created: new ResponseReceipt(),
			current: new ResponseReceipt(),
			pending: new ResponseReceipt(),

			member: new ConsumerMember(),
			principal: new ConsumerPrincipal(),
			preference: new ConsumerPreference(),
			schedule: new ConsumerSchedule(),
			request: new ConsumerRequest(),

			lastActivity: undefined,

			canBecomeAvailable: undefined,
			canBecomePending: undefined,
			canBecomeDismissed: undefined,

			update: function (properties) {
				properties = properties || { };

				// update properties
				this.id = properties.id;
				this.consumerPortfolioId = properties.consumerPortfolioId;
				this.providerPortfolioId = properties.providerPortfolioId;

				this.distance = properties.distance;
				this.quote = properties.quote;
				this.depositDue = properties.depositDue;
				this.depositPaid = properties.depositPaid;
				this.balanceDue = properties.balanceDue;
				this.balancePaid = properties.balancePaid;

				this.agreement.update(properties.agreement);
				this.alternate.update(properties.alternate);

				this.created.update(properties.created);
				this.current.update(properties.current);
				this.pending.update(properties.pending);

				this.member.update(properties.member);
				this.principal.update(properties.principal);
				this.preference.update(properties.preference);
				this.schedule.update(properties.schedule);
				this.request.update(properties.request);

				this.lastActivity = properties.lastActivity;

				this.canBecomeAvailable = properties.canBecomeAvailable;
				this.canBecomePending = properties.canBecomePending;
				this.canBecomeDismissed = properties.canBecomeDismissed;
			},

			toJSON: function () {
				return {
					id: this.id,
					consumerPortfolioId: this.consumerPortfolioId,
					providerPortfolioId: this.providerPortfolioId,
					distance: this.distance,
					quote: this.quote,
					depositDue: this.depositDue,
					depositPaid: this.depositPaid,
					balanceDue: this.balanceDue,
					balancePaid: this.balancePaid,
					agreement: this.agreement.toJSON(),
					alternate: this.alternate.toJSON(),
					current: this.current.toJSON()
				};
			}
		});

		this.update(properties);
	};

 	return ProviderResponse;
}]);

File: Source/Obsequy.Model/Scheme/Enitity/ProviderResponseScheme.cs
using FluentValidation.Results;
using Obsequy.Utility;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MongoDB.Bson.Serialization.Attributes;

namespace Obsequy.Model
{
	public class ProviderResponseScheme
	{
		#region Properties

		public string Id { get; set; }
		public string ConsumerPortfolioId { get; set; }
		public string ProviderPortfolioId { get; set; }

		public double? Distance { get; protected set; }
		public double? Quote { get; set; }
		public double? DepositDue { get; set; }
		public double? DepositPaid { get; set; }
		public double? BalanceDue { get; set; }
		public double? BalancePaid { get; set; }

		public ResponseAgreement Agreement { get; set; }
		public ResponseAlternate Alternate { get; set; }

		public ResponseReceipt Created { get; set; }
		public ResponseReceipt Current { get; set; }
		public ResponseReceipt Pending { get; set; }

		public ConsumerMember Member { get; set; }
		public ConsumerPrincipal Principal { get; set; }

		public ConsumerPreference Preference { get; protected set; }
		public ConsumerSchedule Schedule { get; protected set; }
		public ConsumerRequest Request { get; protected set; }

		public bool CanBecomeAvailable { get; protected set; }
		public bool CanBecomeDismissed { get; protected set; }
		public bool CanBecomePending { get; protected set; }

		#endregion

		#region Computed Properties

		[BsonIgnore]
		public DateTime? LastActivity
		{
			get
			{
				if (Request != null ) {
					// set the last activity to the requests date
					var lastActivity = Request.Current.On;

					if (Current != null && lastActivity < Current.On)
						lastActivity = Current.On;

					if (Pending != null && lastActivity < Pending.On)
						lastActivity = Pending.On;

					return lastActivity;
				}

				return null;
			}
		}

		#endregion

		#region Constructors

		public ProviderResponseScheme()
		{

		}

		public ProviderResponseScheme(Response response, ProviderPortfolio providerPortfolio, ConsumerPortfolio consumerPortfolio, ConsumerMember consumerMember = null)
		{
			this.Id = response.Id;
			this.ConsumerPortfolioId = response.ConsumerPortfolioId;
			this.ProviderPortfolioId = response.ProviderPortfolioId;

			this.Distance = response.Distance;
			this.Quote = response.Quote;
			this.DepositDue = response.DepositDue;
			this.DepositPaid = response.DepositPaid;
			this.BalanceDue = response.BalanceDue;
			this.BalancePaid = response.BalancePaid;

			this.Agreement = response.Agreement;
			this.Alternate = response.Alternate;

			this.Created = response.Created;
			this.Current = response.Current;
			this.Pending = response.Pending;

			this.Member = GetConsumerMember(response, consumerMember); // only show if winning bidder
			this.Principal = (response.Current.IsAccepted ? consumerPortfolio.Principal : new ConsumerPrincipal()); // only show if winning bidder
			this.Preference = consumerPortfolio.Preference.Clone();
			this.Schedule = consumerPortfolio.Schedule.Clone();
			this.Request = consumerPortfolio.Request;

			this.CanBecomeAvailable = response.CanBecomeAvailable;
			this.CanBecomePending = response.CanBecomePending;
			this.CanBecomeDismissed = response.CanBecomeDismissed;
		}

		#endregion

		#region Methods
		private ConsumerMember GetConsumerMember(Response response, ConsumerMember consumerMember)
		{
			var member = new ConsumerMember();

			if (response.Current.IsAccepted && consumerMember != null)
			{
				// only return specific properties
				member.FirstName = consumerMember.FirstName;
				member.LastName = consumerMember.LastName;
				member.Email = consumerMember.Email;
				member.Phone = consumerMember.Phone;
			}

			return member;
		}
		#endregion

		#region Validation
		public ValidationResult Validate(AccountSession accountSession, ValidationMode validationMode)
		{
			return ((new ProviderResponseSchemeValidator(accountSession, validationMode, this)).Validate(this) as ValidationResult);
		}
		#endregion
	}
}
